dockstore-webservice/src/main/java/io/dockstore/webservice/core/SourceFile.java

```diff
@@ -234,6 +239,14 @@ public void setFrozen(boolean frozen) {
         this.frozen = frozen;
     }
 
+    public String getTypeVersion() {
+        return typeVersion;
+    }
+
+    public void setTypeVersion(String typeVersion) {
```

Comment:
As Kathy mentions, CWLs can have mixed versions: https://www.commonwl.org/v1.2/Workflow.html#Syntax
And, since multiple workflows/tools can appear in a single descriptor file (either nested, or in the case that the file describes a number of workflow steps, at the same depth level), the corresponding sourcefile can have multiple language versions.  I'm not sure if it's worth accounting for, though, as it's probably somewhat unusual.

Reply:
So, we understand that a CWL file of one version can reference or use as subworkflows, CWL files of other versions. Are you saying that a CWL file can have multiple versions inside itself? 

That is a horrifying thought. And yeah, I would probably spin off a ticket

Reply:
Yeah, a CWL workflow can be split into files that represent the constituent processes (tools and subworkflows), or all of it can be included in one big file (containing the top level workflow and all of the constituent processes).  Each tool and [sub]workflow can have a different version, and if they're in the same file...

The exception to the above: the spec says that if the CWL is packed (a variant representation where, by definition, all of the parts _are_ in one file) any versions in the constituent processes are to be ignored.
 
Side note: a CWL does not necessarily need to be broken into pieces on the tool/workflow boundaries.  As long as a file is valid YAML, it can be `$import`ed into a CWL and used to represent whatever portion of the CWL is desired (inputs, outputs, workflow steps, etc)...

Reply:
> or all of it can be included in one big file (containing the top level workflow and all of the constituent processes). Each tool and [sub]workflow can have a different version, and if they're in the same file...

gross

Reply:
> the corresponding sourcefile can have multiple language versions

@svonworl are you suggesting that `SourceFile` should store a list of versions like `Version`?

Reply:
> > the corresponding sourcefile can have multiple language versions
> 
> @svonworl are you suggesting that `SourceFile` should store a list of versions like `Version`?

No, I think that might be overkill (but technically, if we were to be 100% consistent with the dark corners of CWL, yes, it probably should.)